src/XML/src/FMI3/fmi3_xml_terminals_and_icons.c

```diff
@@ -0,0 +1,238 @@
+/*
+    Copyright (C) 2023 Modelon AB
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the BSD style license.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    FMILIB_License.txt file for more details.
+
+    You should have received a copy of the FMILIB_License.txt file
+    along with this program. If not, contact Modelon AB <http://www.modelon.com>.
+*/
+
+#include <stdio.h>
+
+
+#include "JM/jm_named_ptr.h"
+#include "fmi3_xml_parser.h"
+#include "fmi3_xml_terminals_and_icons_impl.h"
+
+static const char* module = "FMI3XML";
+
+fmi3_xml_terminals_and_icons_t* fmi3_xml_allocate_terminals_and_icons(jm_callbacks* callbacks) {
+    jm_callbacks* cb = callbacks ? callbacks : jm_get_default_callbacks();
+
+    fmi3_xml_terminals_and_icons_t* termIcon = (fmi3_xml_terminals_and_icons_t*)callbacks->calloc(1, sizeof(fmi3_xml_terminals_and_icons_t));
+    if (!termIcon) {
+        jm_log_fatal(cb, module, "Could not allocate memory");
+        return 0;
+    }
+    termIcon->callbacks = cb;
+    termIcon->md = NULL;
+
+    jm_vector_init(char)(&termIcon->fmi3_xml_standard_version, 0, cb);
+
+    jm_vector_init(jm_voidp)(&termIcon->terminalsOrigOrder, 0, cb);
+    jm_vector_init(jm_named_ptr)(&termIcon->terminalsByName, 0, cb);
+
+    jm_vector_init(jm_string)(&termIcon->names, 0, cb);
+
+    return termIcon;
+}
+
+static fmi3_xml_terminal_t* fmi3_xml_alloc_terminal(fmi3_xml_parser_context_t* context) {
+    fmi3_xml_terminal_t* term = (fmi3_xml_terminal_t*)context->callbacks->calloc(1, sizeof(fmi3_xml_terminal_t));
+    if (!term) {
+        fmi3_xml_parse_fatal(context, "Could not allocate memory");
+        return NULL;
+    }
+    return term;
+}
+
+static void fmi3_xml_free_terminal(jm_callbacks* callbacks, fmi3_xml_terminal_t* term) {
+    callbacks->free(term);
+}
+
+void fmi3_xml_free_terminals_and_icons(fmi3_xml_terminals_and_icons_t* termIcon) {
+    if (!termIcon) {return;}
+    jm_callbacks* callbacks = termIcon->callbacks;
+    if (termIcon) {
+        jm_vector_free_data(char)(&(termIcon->fmi3_xml_standard_version));
+
+        // free terminals
+        for (size_t i = 0; i < jm_vector_get_size(jm_voidp)(&(termIcon->terminalsOrigOrder)); i++) {
+            fmi3_xml_terminal_t* term = (fmi3_xml_terminal_t*) jm_vector_get_item(jm_voidp)(&(termIcon->terminalsOrigOrder), i);
+            fmi3_xml_free_terminal(callbacks, term);
+        }
+        jm_vector_free_data(jm_voidp)(&(termIcon->terminalsOrigOrder));
+        jm_vector_free_data(jm_named_ptr)(&termIcon->terminalsByName);
+
+        void(*cb_free)(const char*) = (void(*)(const char*))callbacks->free;
```

Comment:
What does this do?

Reply:
cb_free is a function pointer to a function with void return and const char* input.  
We assign it the callbacks->free function, with some extra typecasts, since callbacks->free is defined for jm_voidp = void* inputs.

Also added a comment for this.